src/main/java/io/cdap/plugin/successfactors/source/metadata/SuccessFactorsSchemaGenerator.java

```diff
@@ -0,0 +1,770 @@
+    }
+
+    List<SuccessFactorsColumnMetadata> columns = new ArrayList<>();
+
+    List<String> propList = complexType.getPropertyNames();
+    for (String prop : propList) {
+      EdmProperty property = ((EdmProperty) complexType.getProperty(prop));
+      columns.add(buildSuccessFactorsColumnMetadata(namespace, property));
+    }
+
+    return columns;
+  }
+
+  /**
+   * Builds schema from the given list of {@code SuccessFactorsColumnMetadata}
+   *
+   * @param columnDetailList {@code SuccessFactorsColumnMetadata}
+   * @return {@code Schema}
+   */
+  private Schema buildSchema(List<SuccessFactorsColumnMetadata> columnDetailList) {
+    List<Schema.Field> outputSchema = columnDetailList.stream()
+      .map(this::buildSchemaField)
+      .collect(Collectors.toList());
+
+    return Schema.recordOf("SuccessFactorsColumnMetadata", outputSchema);
+  }
+
+  /**
+   * Builds Schema field from {@code SuccessFactorsColumnMetadata}
+   *
+   * @param successFactorsColumnDetail {@code SuccessFactorsColumnMetadata}
+   * @return {@code Schema.Field}
+   */
+  private Schema.Field buildSchemaField(SuccessFactorsColumnMetadata successFactorsColumnDetail) {
+    if (successFactorsColumnDetail.containsChild()) {
+      List<Schema.Field> outputSchema = successFactorsColumnDetail.getChildList()
+        .stream()
+        .map(this::buildSchemaField)
+        .collect(Collectors.toList());
+
+      String typeName = buildNestedTypeUniqueName(successFactorsColumnDetail.getName());
+
+      if (successFactorsColumnDetail.getType().equals(EdmTypeKind.ENTITY.name()) &&
+        (successFactorsColumnDetail.getMultiplicityOrdinal() != null &&
+          successFactorsColumnDetail.getMultiplicityOrdinal() > 0)) {
+        // adding 1 to * multiplicity record to ARRAY type
+        return Schema.Field.of(successFactorsColumnDetail.getName(),
+                               Schema.arrayOf(Schema.recordOf(typeName, outputSchema)));
+      }
+
+      // adding 0 to 1 multiplicity record to NULLABLE
+      return Schema.Field.of(successFactorsColumnDetail.getName(),
+                             Schema.nullableOf(Schema.recordOf(typeName, outputSchema)));
+    }
+
+    return Schema.Field.of(successFactorsColumnDetail.getName(), buildRequiredSchemaType(successFactorsColumnDetail));
+  }
+
+  /**
+   * Build and returns the appropriate schema type.
+   *
+   * @param successFactorsColumnDetail {@code SuccessFactorsColumnMetadata}
+   * @return {@code Schema}
+   */
+  private Schema buildRequiredSchemaType(SuccessFactorsColumnMetadata successFactorsColumnDetail) {
+    Schema schemaType = SCHEMA_TYPE_MAPPING.get(successFactorsColumnDetail.getType());
+
+    if (schemaType.getLogicalType() == Schema.LogicalType.DECIMAL
+      && successFactorsColumnDetail.getPrecision() != null
+      && successFactorsColumnDetail.getScale() != null) {
+
+      schemaType = Schema.decimalOf(successFactorsColumnDetail.getPrecision(), successFactorsColumnDetail.getScale());
+    }
```

Comment:
If precision and scale are not specified in successFactorsColumnDetail, then we default to (2, 1)? Can that happen? Should we throw an exception if precision/scale are not available?